parprouted.8
.\" Automatically generated by Pod::Man v1.34, Pod::Parser v1.13
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sh \" Subsection heading
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  | will give a
.\" real vertical bar.  \*(C+ will give a nicer C++.  Capital omega is used to
.\" do unbreakable dashes and therefore won't be available.  \*(C` and \*(C'
.\" expand to `' in nroff, nothing in troff, for use with C<>.
.tr \(*W-|\(bv\*(Tr
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.Sh), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.if \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.\"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.hy 0
.if n .na
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "PARPROUTED 8"
.TH PARPROUTED 8 "January 2004" "parprouted" "Proxy ARP Bridging Daemon"
.SH "NAME"
\&\fBparprouted\fR \- Proxy ARP IP bridging daemon
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBparprouted\fR [\fB\-d\fR] [\fB\-P\fR | \fB--no-proxy\fR] [\fB\-R\fR | \fB--no-routes\fR] \fBinterface0\fR [\fBinterface\fR...]
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
parprouted is a daemon for transparent \s-1IP\s0 (Layer 3) proxy \s-1ARP\s0 bridging and routing.
Unlike standard bridging, proxy \s-1ARP\s0 bridging allows to bridge Ethernet 
networks behind wireless nodes. Normal L2 bridging does not work between 
wireless nodes because wireless does not know about \s-1MAC\s0 addresses used 
in the wired Ethernet networks. Also this daemon is useful
for making transparent firewalls.
.PP
The daemon listens for incoming \s-1ARP\s0 requests and forwards them to other interfaces.
It also listens for \s-1ARP\s0 replies, and in case the reply matches an outstanding
forwarded request then it forwards the reply, using the local hardware address as the source,
which causes the stations to send traffic to the host that the daemon is running on.
.PP
If a host is discovered to be present on \fBinterface0\fR,
by virtue of inspecting received ARP request and reply packets,
then it adds an appropriate /32 route to the Linux kernel \s-1IP\s0 routing 
table.
If a host is discovered to be present on another interface then any route on \fBinterface0\fR
is removed.
All routes entered by the daemon have a metric of 50. 
.PP
For hosts for which routes have been established the daemon probes them every 50 seconds
using ARP. The routes for unresponsive hosts are timed out and removed.
.SH "REQUIREMENTS"
.IX Header "REQUIREMENTS"
This daemon can be used for unicast traffic only. I.e., \s-1DHCP\s0 is not supported.
Use software such as dhcrelay to \*(L"bridge\*(R" \s-1DHCP\s0.
.PP
You should have an \s-1IP\s0 address assigned on all of your interfaces that you do 
bridging on. There is no need for this address to be from the same networks as 
the bridged networks. Any dummy address will do.
.PP
The system should have correct default route.
.PP
parprouted requires \*(L"ip\*(R" program from iproute2 tools to be installed in 
/sbin. If it is installed in another location, please replace \*(L"/sbin/ip\*(R" 
occurances in the source with the correct path.
.PP
parprouted is designed for and tested only with Linux 2.4.x kernels.
.SH "OPTIONS"
.IX Header "OPTIONS"
The list of interfaces to do bridging on should be given via the command
line.
.PP
The daemon accepts the following switches:
.PP
\&\fB\-d\fR, which stands for debugging. If you run it in debugging mode the daemon 
will not go to background and will print additional debugging information to 
stdout/stderr.
.PP
\&\fB\-P\fR or \fB\--no-proxy\fR, disables the proxy-ARP functionality.
.PP
\&\fB\-R\fR or \fB\--no-routes\fR, disables the route-management functionality.
.SH "EXAMPLE"
.IX Header "EXAMPLE"
To bridge between wlan0 and eth0 where wlan0 is the main interface with subnet and/or default routes: \fBparprouted eth0 wlan0\fR
.SH "AUTHOR"
.IX Header "AUTHOR"
.Vb 2
\& (C) 2004, Vladimir Ivaschenko <vi@maks.net>
\& http://www.hazard.maks.net
\& Copyright (C) 2019 Lenbrook Industries Limited
.Ve


parprouted.c
/* parprouted: ProxyARP routing daemon. 
 * (C) 2008 Vladimir Ivaschenko <vi@maks.net>
 * Copyright (C) 2019 Lenbrook Industries Limited
 *
 * This application is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Library General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License.
 *
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	 See the GNU
 * Library General Public License for more details.
 *
 * You should have received a copy of the GNU Library General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
 * Boston, MA 02111-1307, USA.
 *
 */
 
#include "parprouted.h"

char *progname;
int debug=0;
int verbose=0;
int g_perform_shutdown=0;
int g_manage_routes = 1;
int g_proxy_arp = 1;


char *errstr;

pthread_t my_threads[MAX_IFACES+1];
int last_thread_idx=-1;

char * g_ifaces[MAX_IFACES];
int g_last_iface_idx=-1;

typedef struct route_entry {
    struct in_addr ipaddr_ia;
    int idx;
    time_t tstamp;
    int route_added;
    int want_route;
    struct route_entry *next;
} ROUTE_ENTRY;

ROUTE_ENTRY **route_table;
pthread_mutex_t route_table_mutex;

ROUTE_ENTRY * get_route_entry_LOCKED(ROUTE_ENTRY ** route_table_p, struct in_addr ipaddr, int idx)
{
    ROUTE_ENTRY * cur_entry=*route_table_p;
    ROUTE_ENTRY * prev_entry=NULL;

    while (cur_entry != NULL && ( ipaddr.s_addr != cur_entry->ipaddr_ia.s_addr || cur_entry->idx != idx ) ) {
	prev_entry = cur_entry;
	cur_entry = cur_entry->next;
    };

    if (cur_entry == NULL) {
	if (debug) printf("Creating new route_table entry %s(%s)\n", inet_ntoa(ipaddr), g_ifaces[idx]);

	if ((cur_entry = (ROUTE_ENTRY *) malloc(sizeof(ROUTE_ENTRY))) == NULL) {
	    errstr = strerror(errno);
	    syslog(LOG_ERR, "No memory: %s", errstr);
	} else {
	    if (prev_entry == NULL) { *route_table_p=cur_entry; }
	    else { prev_entry->next = cur_entry; }
	    cur_entry->next = NULL;
	    cur_entry->idx = idx;
	    cur_entry->ipaddr_ia.s_addr = ipaddr.s_addr;
	    cur_entry->route_added = 0;
	}
    }
    
    return cur_entry;
}

/* Remove all entires in route_table where ipaddr is NOT on interface dev */
int remove_other_routes_LOCKED(ROUTE_ENTRY * cur_entry, struct in_addr ipaddr, int idx)
{
    int removed = 0;

    for (; cur_entry != NULL; cur_entry = cur_entry->next) {
        if (ipaddr.s_addr == cur_entry->ipaddr_ia.s_addr && idx != cur_entry->idx)  {
            if (debug && cur_entry->want_route) printf("Marking entry %s(%s) for removal\n", inet_ntoa(ipaddr), g_ifaces[cur_entry->idx]);
            cur_entry->want_route = 0;
            ++removed;
        }
    }
    if (debug && removed)
	printf("Found entry %s(%s), removed entries via other interfaces\n",
		inet_ntoa(ipaddr), g_ifaces[idx]);

    return removed;
}

/* Remove route from kernel */
int route_remove_LOCKED(ROUTE_ENTRY* entry)
{
    char routecmd_str[ROUTE_CMD_LEN];
    int success = 1;
    
    if (snprintf(routecmd_str, ROUTE_CMD_LEN-1, 
	    "/sbin/ip route del %s/32 metric 50 dev %s scope link",
	    inet_ntoa(entry->ipaddr_ia), g_ifaces[entry->idx]) > ROUTE_CMD_LEN-1)
    {
	syslog(LOG_ERR, "ip route command too large to fit in buffer!");
    } else {
	if (system(routecmd_str) != 0)
	{
	    syslog(LOG_ERR, "'%s' unsuccessful!", routecmd_str);
	    if (debug) printf("%s failed\n", routecmd_str);
	    success = 0;
	}
	else 
	{
	    syslog(LOG_INFO, "route added for %s via %s", inet_ntoa(entry->ipaddr_ia), g_ifaces[entry->idx]);
	    if (debug) printf("%s success\n", routecmd_str);
	    success = 1;
	}
    }
    if (success)
	entry->route_added = 0;
    
    return success;
}

/* Add route into kernel */
int route_add_LOCKED(ROUTE_ENTRY* entry)
{
    char routecmd_str[ROUTE_CMD_LEN];
    int success = 1;

    if (snprintf(routecmd_str, ROUTE_CMD_LEN-1, 
	    "/sbin/ip route add %s/32 metric 50 dev %s scope link",
	    inet_ntoa(entry->ipaddr_ia), g_ifaces[entry->idx]) > ROUTE_CMD_LEN-1)
    {
	syslog(LOG_ERR, "ip route command too large to fit in buffer!");
    } else {
	if (system(routecmd_str) != 0)
	{ 
	    syslog(LOG_ERR, "'%s' unsuccessful, will try to remove!", routecmd_str);
	    if (debug) printf("%s failed\n", routecmd_str);
	    route_remove_LOCKED(entry);
	    success = 0;
	}
	else
	{
	    syslog(LOG_INFO, "route removed for %s via %s", inet_ntoa(entry->ipaddr_ia), g_ifaces[entry->idx]);
	    if (debug) printf("%s success\n", routecmd_str);
	    success = 1;
	}
    }
    if (success)
	entry->route_added = 1;

    return success;
}


void manage_route_entries_LOCKED(ROUTE_ENTRY **route_table_p, int in_cleanup)
{
    ROUTE_ENTRY *cur_entry=*route_table_p, *prev_entry=NULL;

    /* First loop to remove unwanted routes */
    while (cur_entry != NULL) {
	if (debug && verbose) printf("Working on route %s(%s) tstamp %u want_route %d\n", inet_ntoa(cur_entry->ipaddr_ia), g_ifaces[cur_entry->idx], (int) cur_entry->tstamp, cur_entry->want_route);

	if ( !cur_entry->want_route
	    || time(NULL) - cur_entry->tstamp > ARP_TABLE_ENTRY_TIMEOUT 
	    || in_cleanup)  {
	    
	    if (cur_entry->route_added)
		route_remove_LOCKED(cur_entry);

	    /* remove from arp list */
	    if (debug) printf("Delete arp %s(%s)\n", inet_ntoa(cur_entry->ipaddr_ia), g_ifaces[cur_entry->idx]);
		
	    if (prev_entry != NULL) {
	        prev_entry->next = cur_entry->next;
	        free(cur_entry);
	        cur_entry=prev_entry->next;
	    } else {
	        *route_table_p = cur_entry->next;
	        free(cur_entry);
        	cur_entry=*route_table_p;
            }
	} else {
    	    prev_entry = cur_entry;
	    cur_entry = cur_entry->next;
	}	
    } /* while loop */

    /* Now loop to add new routes */
    cur_entry=*route_table_p;
    while (cur_entry != NULL) {
	if (time(NULL) - cur_entry->tstamp <= ARP_TABLE_ENTRY_TIMEOUT 
		&& cur_entry->want_route
		&& !cur_entry->route_added 
		&& !in_cleanup) 
	{
	    /* add route to the kernel */
	    route_add_LOCKED(cur_entry);
	}
	cur_entry = cur_entry->next;
    } /* while loop */

}	

int add_route_LOCKED(ROUTE_ENTRY **route_table_p, const struct in_addr in, const int idx)
{
    ROUTE_ENTRY *entry;
    int added;

    entry = get_route_entry_LOCKED(route_table_p, in, idx);

    added = !entry->want_route;

    entry->want_route = 1;

    time(&entry->tstamp);

    manage_route_entries_LOCKED(route_table_p, 0);

    if (debug && (!entry->route_added || added)) {
	printf("route_table entry: '%s' Dev: '%s' route_added:%d aded:%d\n",
	        inet_ntoa(entry->ipaddr_ia),
	        g_ifaces[entry->idx], entry->route_added,
	        added);
    }

    return added;
}

void manage_route(const struct in_addr address, const int idx)
{
    int manage_routes;

    assert(idx <= g_last_iface_idx);

    if (!g_manage_routes) return;

    pthread_mutex_lock(&route_table_mutex);
    manage_routes = remove_other_routes_LOCKED(*route_table, address, idx);
    if (idx == 0) {
	manage_routes += add_route_LOCKED(route_table, address, idx);
    }
    if (manage_routes) manage_route_entries_LOCKED(route_table, 0);
    pthread_mutex_unlock(&route_table_mutex);
}

void remove_routes(int idx)
{
    int removed = 0;
    ROUTE_ENTRY *cur_entry;

    assert(idx <= g_last_iface_idx);

    pthread_mutex_lock(&route_table_mutex);

    for (cur_entry = *route_table; cur_entry != NULL; cur_entry = cur_entry->next) {
	if (cur_entry->idx == idx && cur_entry->want_route) {
	    cur_entry->want_route = 1;
	    removed++;
	}
    }
    if (removed) manage_route_entries_LOCKED(route_table, 0);

    pthread_mutex_unlock(&route_table_mutex);

    if (debug && removed)
	printf("%s, removed route entries for interface\n", g_ifaces[idx]);

}


/* ARP ping all entries in the table */

void refresharp_LOCKED(ROUTE_ENTRY *list)
{
  if (debug)
      printf("Refreshing ARP entries.\n");

  while(list != NULL) {
    arp_req(g_ifaces[list->idx], list->ipaddr_ia, 0);
    list = list->next;
  }
}

void cleanup() 
{
    int i;
    
    syslog(LOG_INFO, "Received signal; cleaning up.");

    for (i=0; i <= last_thread_idx; i++) {
	void * ret;
	pthread_kill(my_threads[i], SIGTERM);
	pthread_join(my_threads[i], &ret);
    }

    pthread_mutex_trylock(&route_table_mutex);
    manage_route_entries_LOCKED(route_table, 1);
    pthread_mutex_unlock(&route_table_mutex);

    syslog(LOG_INFO, "Terminating.");
}

void sighandler()
{
    g_perform_shutdown=1;
}

void *main_thread()
{
    time_t last_refresh;

    while (!g_perform_shutdown) {
        pthread_mutex_lock(&route_table_mutex);
        manage_route_entries_LOCKED(route_table, 0);
	pthread_mutex_unlock(&route_table_mutex);
	if (g_perform_shutdown) break;
	usleep(SLEEPTIME);
	if (g_perform_shutdown) break;
	if (time(NULL)-last_refresh > REFRESHTIME) {
	    pthread_mutex_lock(&route_table_mutex);
	    refresharp_LOCKED(*route_table);
	    pthread_mutex_unlock(&route_table_mutex);
	    time(&last_refresh);
	}
    }
    return 0;
}
    
int main (int argc, char **argv)
{
    pid_t child_pid;
    int i, help=1;
    struct sigaction handle;

    
    progname = (char *) basename(argv[0]);
    
    for (i = 1; i < argc; i++) {
	if (!strcmp(argv[i],"-d")) { 
	    debug=1;
	    help=0;
	}
	else if (!strcmp(argv[i],"-P") || !strcmp(argv[i],"--no-proxy")) {
	    g_proxy_arp = 0;
	}
	else if (!strcmp(argv[i],"-R") || !strcmp(argv[i],"--no-routes")) {
	    g_manage_routes = 0;
	}
	else if (!strcmp(argv[i],"-h") || !strcmp(argv[i],"--help")) {
	    help = 1;
	    break;
	}
	else {
	    g_last_iface_idx++;
	    g_ifaces[g_last_iface_idx]=argv[i];
	    help=0;
	}
    }

    if (help || g_last_iface_idx <= -1) {
	    printf("parprouted: proxy ARP routing daemon, version %s.\n", VERSION);
    	    printf("(C) 2007 Vladimir Ivaschenko <vi@maks.net>, GPL2 license.\n");
    	    printf("(C) 2019 Lenbrook Industries Limited.\n");
	    printf("Usage: parprouted [-d] [-P | --no-proxy] [-R | --no-routes] interface [interface]\n");
	    exit(1);
    }

    if (!debug) {
        /* fork to go into the background */
        if ((child_pid = fork()) < 0) {
            fprintf(stderr, "could not fork(): %s", strerror(errno));
            exit(1);
        } else if (child_pid > 0) {
            /* fork was ok, wait for child to exit */
            if (waitpid(child_pid, NULL, 0) != child_pid) {
                perror(progname);
                exit(1);
            }
            /* and exit myself */
            exit(0);
        }
        /* and fork again to make sure we inherit all rights from init */
        if ((child_pid = fork()) < 0) {
            perror(progname);
            exit(1);
        } else if (child_pid > 0)
            exit(0);

        /* create our own session */
        setsid();

        /* close stdin/stdout/stderr */
        close(0);
        close(1);
        close(2);

    }

    openlog(progname, LOG_PID | LOG_CONS | LOG_PERROR, LOG_DAEMON);
    syslog(LOG_INFO, "Starting");

    memset(&handle, 0, sizeof(handle));
    handle.sa_handler   = sighandler;
    handle.sa_flags     = 0; // !SA_RESTART

    sigaction(SIGINT, &handle, 0);
    sigaction(SIGTERM, &handle, 0);
    sigaction(SIGHUP, &handle, 0);

    if ((route_table = (ROUTE_ENTRY **) malloc(sizeof(ROUTE_ENTRY **))) == NULL) {
	    errstr = strerror(errno);
	    syslog(LOG_CRIT, "No memory: %s", errstr);
	    abort();
    }
    
    *route_table = NULL;

    pthread_mutex_init(&route_table_mutex, NULL);
    
    for (i=0; i <= g_last_iface_idx; i++) {
	int * idx = malloc(sizeof(idx));
	*idx = i;
	if (pthread_create(&my_threads[++last_thread_idx], NULL, arp, idx)) {
	    syslog(LOG_CRIT, "Error creating ARP thread for %s",g_ifaces[i]);
	    abort();
	}
	if (debug) printf("Created ARP thread for %s.\n",g_ifaces[i]);
    }

    main_thread();

    cleanup();

    return 0;
}
